File: src/java_practice/ComparatorPractice1.java
package java_practice;

import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import org.jetbrains.annotations.NotNull;

public class ComparatorPractice1 {
  private static final Comparator<Student> comparator1 =
      Comparator.comparing(Student::getBirth)
          .thenComparing(Comparator.comparing(Student::getName));

  public static void main(String[] args){
    Student student1 = new Student("1990-01-01", "John");
    Student student2 = new Student("1990-01-02", "Jane");
    Student student3 = new Student("1990-01-02", "Jane2");

    List<Student> students = Arrays.asList(student1, student2, student3);

//    students.sort(comparator1);
    students.sort(Student::compareTo);

    students.forEach(System.out::println);
  }
}

class Student implements Comparable<Student>{
  private final String birth;
  private final String name;

  Student(String birth, String name) {
    this.birth = birth;
    this.name = name;
  }

  public String getName() {
    return name;
  }

  public String getBirth() {
    return birth;
  }

  @Override
  public String toString() {
    return "Student{" +
        "birth='" + birth + '\'' +
        ", name='" + name + '\'' +
        '}';
  }

  @Override
  public int compareTo(@NotNull Student other) {
    int result = this.birth.compareTo(other.birth);
    if(result != 0) {
      return result;
    }

    result = other.name.compareTo(this.name);
    if(result != 0) {
      return result;
    }

    return 0;
  }
}

File: src/KotlinPractice/ComparatorPractice1.kt
package KotlinPractice

data class StudentKt(
    val birth: String,
    val name: String
) : Comparable<StudentKt> {

    override fun compareTo(other: StudentKt): Int {
        val result = this.birth.compareTo(other.birth)
        if (result != 0) return result

        return other.name.compareTo(this.name)
    }
}

fun main() {
    val comparator1 = compareBy(StudentKt::birth)
        .thenBy { it.name }

    val student1 = StudentKt("1990-01-01", "John")
    val student2 = StudentKt("1990-01-02", "Jane")
    val student3 = StudentKt("1990-01-02", "Jane2")

    val students = mutableListOf(student1, student2, student3)

//    students.sortWith(comparator1)
    students.sort()

    students.forEach(::println)
}
